components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -1826,97 +1821,67 @@ public int compare(Tag o1, Tag o2) {
-            //as a tenant, I'm browsing my own Store or I'm browsing a Store of another tenant..
-            if ((this.isTenantModeStoreView && this.tenantDomain==null) || (this.isTenantModeStoreView && isTenantDomainNotMatching(requestedTenantDomain))) {//Tenant based store anonymous mode
-                int tenantId = getTenantId(this.requestedTenant);
-                userRegistry = ServiceReferenceHolder.getInstance().getRegistryService().
-                        getGovernanceUserRegistry(CarbonConstants.REGISTRY_ANONNYMOUS_USERNAME, tenantId);
-            } else {
-                userRegistry = registry;
-            }
-
-            Map<String, Tag> tagsData = new HashMap<String, Tag>();
-            try {
-            	PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(((UserRegistry)userRegistry).getUserName());
-                if (requestedTenant != null ) {
-                    isTenantFlowStarted = startTenantFlowForTenantDomain(requestedTenant);
-                    PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(((UserRegistry)userRegistry).getUserName());
-                }
-
-                Map <String, List<String>> criteriaPublished = new HashMap<String, List<String>>();
-                criteriaPublished.put(APIConstants.LCSTATE_SEARCH_KEY, new ArrayList<String>() {{
-                    add(APIConstants.PUBLISHED);
-                }});
-                //rxt api media type
-                List<TermData> termsPublished = GovernanceUtils
-                        .getTermDataList(criteriaPublished, APIConstants.API_OVERVIEW_TAG,
-                                         APIConstants.API_RXT_MEDIA_TYPE, true);
-
-                if(termsPublished != null){
-                    for(TermData data : termsPublished){
-                        tempTagSet.add(new Tag(data.getTerm(), (int)data.getFrequency()));
-                    }
-                }
-
-                Map<String, List<String>> criteriaPrototyped = new HashMap<String, List<String>>();
-                criteriaPrototyped.put(APIConstants.LCSTATE_SEARCH_KEY, new ArrayList<String>() {{
-                    add(APIConstants.PROTOTYPED);
-                }});
-                //rxt api media type
-                List<TermData> termsPrototyped = GovernanceUtils
-                        .getTermDataList(criteriaPrototyped, APIConstants.API_OVERVIEW_TAG,
-                                         APIConstants.API_RXT_MEDIA_TYPE, true);
-
-                if(termsPrototyped != null){
-                    for(TermData data : termsPrototyped){
-                        tempTagSet.add(new Tag(data.getTerm(), (int)data.getFrequency()));
+            DevPortalAPISearchResult searchAPIs = apiPersistenceInstance.searchAPIsForDevPortal(org, "",
+                    0, 25, userCtx);
+            SortedSet<Object> apiSet = new TreeSet<>(new APIAPIProductNameComparator());
```

Comment:
Same here.

Reply:
we can easily do an aggregation query and fetch all the tags and do a search, are we going to return an API after searching by the tags ? if so we do not need a new collection

Reply:
We only need tags and API categories. Not returning an API.

Reply:
The aggregation query anyway first needs to get all the matching APIs that need to be displayed in the dev portal. And then only we can retrieve the tags from those APIs. Thus we need to traverse all the APIs
( Displaying the tags of Deprecated and Retired APIs won't make the sense Nah?)

Reply:
Implemented Approach 2 according to the offline discussion

https://github.com/wso2/carbon-apimgt/pull/10139/commits/c20a95b3628c2dfe2724d9f0c808258afea566b6